pkg/controller/baremetalhost/host_state_machine.go

```diff
@@ -0,0 +1,294 @@
+	return map[metal3v1alpha1.ProvisioningState]stateHandler{
+		metal3v1alpha1.StateNone:                  hsm.handleNone,
+		metal3v1alpha1.StateRegistering:           hsm.handleRegistering,
+		metal3v1alpha1.StateRegistrationError:     hsm.handleRegistrationError,
+		metal3v1alpha1.StateInspecting:            hsm.handleInspecting,
+		metal3v1alpha1.StateExternallyProvisioned: hsm.handleExternallyProvisioned,
+		metal3v1alpha1.StateMatchProfile:          hsm.handleMatchProfile,
+		metal3v1alpha1.StateReady:                 hsm.handleReady,
+		metal3v1alpha1.StateProvisioning:          hsm.handleProvisioning,
+		metal3v1alpha1.StateProvisioningError:     hsm.handleProvisioningError,
+		metal3v1alpha1.StateProvisioned:           hsm.handleProvisioned,
+		metal3v1alpha1.StatePowerManagementError:  hsm.handlePowerManagementError,
+		metal3v1alpha1.StateDeprovisioning:        hsm.handleDeprovisioning,
+		metal3v1alpha1.StateDeleting:              hsm.handleDeleting,
+	}
+}
+
+func (hsm *hostStateMachine) updateHostStateFrom(initialState metal3v1alpha1.ProvisioningState,
+	result *reconcile.Result,
+	log logr.Logger) {
+	if hsm.NextState != initialState {
+		log.Info("changing provisioning state",
+			"old", initialState,
+			"new", hsm.NextState)
+		hsm.Host.Status.Provisioning.State = hsm.NextState
+		result.Requeue = true
+	}
+}
+
+func (hsm *hostStateMachine) ReconcileState(info *reconcileInfo) (result reconcile.Result, err error) {
+	initialState := hsm.Host.Status.Provisioning.State
+	defer hsm.updateHostStateFrom(initialState, &result, info.log)
+
+	if hsm.shouldInitiateDelete() {
+		info.log.Info("Initiating host deletion")
+		return
+	}
+	// TODO: In future we should always re-register the host if required,
+	// rather than initiate a transistion back to the Registering state.
+	if hsm.shouldInitiateRegister(info) {
+		info.log.Info("Initiating host registration")
+		return
+	}
+
+	if stateHandler, found := hsm.handlers()[initialState]; found {
+		result, err = stateHandler(info)
+	} else {
+		info.log.Info("No handler found for state", "state", initialState)
+		err = fmt.Errorf("No handler found for state \"%s\"", initialState)
+		return
+	}
+
+	return
+}
+
+func (hsm *hostStateMachine) shouldInitiateDelete() bool {
+	if hsm.Host.DeletionTimestamp.IsZero() {
+		// Delete not requested
+		return false
+	}
+
+	hsm.RequeueDespiteError = true
+	switch hsm.NextState {
+	default:
+		hsm.NextState = metal3v1alpha1.StateDeleting
+	case metal3v1alpha1.StateProvisioning, metal3v1alpha1.StateProvisioningError, metal3v1alpha1.StateProvisioned:
+		hsm.NextState = metal3v1alpha1.StateDeprovisioning
+	case metal3v1alpha1.StateDeprovisioning:
+		// Allow state machine to run to continue deprovisioning.
+		return false
+	case metal3v1alpha1.StateDeleting:
+		// Already in deleting state. Allow state machine to run.
+		return false
+	}
+	return true
+}
+
+func (hsm *hostStateMachine) shouldInitiateRegister(info *reconcileInfo) bool {
```

Comment:
The same comment about side-effects applies here, too.

Reply:
I have an upcoming [patch](https://github.com/zaneb/baremetal-operator/commit/a65c1d107dece244490157db31dcfc76882fd8a6) that replaces this function with one called `ensureRegistered`, so that should resolve this.